### PR TITLE
bpf: unit tests and fixes for `prepend_name` function

### PR DESCRIPTION
**Pull request by mtardy**
See commits!

### CODE REVIEW

File: bpf/process/bpf_process_event.h [hunk 1]
```diff
@@ -108,42 +114,50 @@ d_unlinked(struct dentry *dentry)
...
 	// Needed to bound that for probe_read call.
-	if (buffer_offset > PATH_MAP_SIZE - 256)
+	if (buffer_offset >= MAX_BUF_LEN)
 		return -ENAMETOOLONG;
 
-	probe_read(bf + buffer_offset, sizeof(char), &slash);
+	if (write_slash)
+		probe_read(buf + buffer_offset, sizeof(char), &slash);
+
 	// This ensures that dlen is < 256, which is aligned with kernel's max dentry name length
```
Comment (willfindlay) [inline bpf/process/bpf_process_event.h:152]: dlen -> namelen to match the variable rename

File: bpf/process/bpf_process_event.h [hunk 2]
```diff
@@ -108,42 +114,50 @@ d_unlinked(struct dentry *dentry)
...
 	// with the current buffer pointer. This will be at max 256 bytes (similar to the initial
 	// size).
 	// Needed to bound that for probe_read call.
-	if (buffer_offset > PATH_MAP_SIZE - 256)
+	if (buffer_offset >= MAX_BUF_LEN)
 		return -ENAMETOOLONG;
 
-	probe_read(bf + buffer_offset, sizeof(char), &slash);
+	if (write_slash)
+		probe_read(buf + buffer_offset, sizeof(char), &slash);
```
Comment (willfindlay) [inline bpf/process/bpf_process_event.h:150]: Hm, couldn't this be simplified to avoid the probe_read call?

File: bpf/tests/prepend_name_test.c [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+};
+
+struct {
+	__uint(type, BPF_MAP_TYPE_ARRAY);
+	__uint(max_entries, 1);
+	__type(key, int);
+	__type(value, struct test_prepend_name_state_map_value);
+} test_prepend_name_state_map SEC(".maps");
+
+__attribute__((section("raw_tracepoint/test"), used)) int
```
Comment (willfindlay) [inline bpf/tests/prepend_name_test.c:34]: This BPF unit testing is extremely valuable and honestly long overdue. Thanks for adding the first of hopefully many to come

**Review by willfindlay — APPROVED** [review]
Just a couple nits, but LGTM. Might be helpful to have someone else do a pass of the `prepend_name` changes, but I didn't find any obvious bugs.

File: bpf/process/bpf_process_event.h (hunk 2 above)
Comment (mtardy) [reply bpf/process/bpf_process_event.h:150]: This works on `6.5.0`, let's see if it passes tests on older versions.

```diff
diff --git a/bpf/process/bpf_process_event.h b/bpf/process/bpf_process_event.h
index ff145373..b2b46845 100644
--- a/bpf/process/bpf_process_event.h
+++ b/bpf/process/bpf_process_event.h
@@ -116,7 +116,6 @@ d_unlinked(struct dentry *dentry)
 static inline __attribute__((always_inline)) int
 prepend_name(char *buf, char **bufptr, int *buflen, const char *name, u32 namelen)
 {
-       char slash = '/';
        // contains 1 if the buffer is large enough to contain the whole name and a slash prefix
        bool write_slash = 1;
 
@@ -147,7 +146,7 @@ prepend_name(char *buf, char **bufptr, int *buflen, const char *name, u32 namele
                return -ENAMETOOLONG;
 
        if (write_slash)
-               probe_read(buf + buffer_offset, sizeof(char), &slash);
+               buf[buffer_offset] = '/';
 
        // This ensures that namelen is < 256, which is aligned with kernel's max dentry name length
        // that is 255 (https://elixir.bootlin.com/linux/v5.10/source/include/uapi/linux/limits.h#L12).
```

**Comment by mtardy** [comment]
Note that this has a chance of fixing https://github.com/cilium/tetragon/issues/1707, it can be related.

File: bpf/tests/prepend_name_test.c [hunk 4]
```diff
@@ -0,0 +1,56 @@
+// SPDX-License-Identifier: (GPL-2.0-only OR BSD-2-Clause)
+/* Copyright Authors of Cilium */
+
+//go:build ignore
```
Comment (tpapagian) [inline bpf/tests/prepend_name_test.c:4]: Question: Why do we need that?

**Review by tpapagian — APPROVED** [review]
LGTM, thanks! Having a skeleton to build eBPF unit tests would be really useful!

File: bpf/tests/prepend_name_test.c (hunk 4 above)
Comment (mtardy) [reply bpf/tests/prepend_name_test.c:4]: So this is a bit weird but you can't have a `.c` file in the same directory as a go module unless you are using CGO 🤔  most Go commands will return `C source files not allowed when not using cgo`.

Comment (mtardy) [reply bpf/tests/prepend_name_test.c:4]: I wanted to house the C file with the BPF test harness and the Go file in the same directory so it's needed!

**Comment by kkourt** [comment]
🚀 

@mtardy I feel like we should backport this change to 1.0. WDYT?

**Comment by mtardy** [comment]
> 🚀 
> 
> 
> 
> @mtardy I feel like we should backport this (or at least the buffix) to 1.0. WDYT?

Yes, it would make sense since it might fix a user facing bug. Let me do that!

**Pull request merged**